### PR TITLE
[IA-3747][IA-3744] Refactor utilities to use typescript

### PR DESCRIPTION
**Pull request by jdcanas**
This PR is an incremental piece of the larger PR to add typescript to the analysis duplicator, [here](https://github.com/DataBiosphere/terra-ui/pull/3447/files). After some feedback and discussion, we decided to split the original PR into 3 parts.

1. Refactor ajax modules to be better compatible with typescript and not have circular dependencies, PR [here](https://github.com/DataBiosphere/terra-ui/pull/3483)  (thanks Nick!)
2. Introduce types into the base utility files that subsequent typescript PRs will depend on (this PR)
3. Type the AnalysisDuplicator component and add typed tests. (This will be rebased off of dev once the previous 2 PRs are merged to preserve all downstream changes). 

Note that this PR also creates the AnalysisDuplicator file, but leaves it identical to the original util file implementation. This involves a lot of `ts-expect-error` annotations to minimize changes not relevant to the goal of this PR. These annotations are removed in the linked PR with proper typescript implementations.

### CODE REVIEW

File: src/libs/testing/test-utils.ts
Comment (nawatts) [inline src/libs/testing/test-utils.ts]: This file already exists as [src/test-utils.ts](https://github.com/DataBiosphere/terra-ui/blob/dev/src/test-utils.ts). It can be moved, but lets not create multiple copies of it.

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/file-utils.ts]: src/libs/type-utils is a good place for general helpers such as NominalType.

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/file-utils.ts]: Maybe call this "AnalysisFileMetadata", since there are other types of files that don't have this metadata.

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/file-utils.ts]: Why does this need a global store vs storing this information in a state variable?

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/file-utils.ts]: There's a type in [src/libs/type-utils/LoadedState.ts](https://github.com/DataBiosphere/terra-ui/blob/dev/src/libs/type-utils/LoadedState.ts) with a pattern for representing the state of fetching some data.

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/file-utils.ts]: Since the `else` block is very short, to reduce indenting, this function could be reordered like:

```js
if (!canShare) {
  return {}
}

...
```

https://medium.com/@timothydan/javascript-guard-clauses-64b999e3240

File: src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts:22]: When variables are initialized with an empty array, TS can't infer what the element type should be. But you can pass a type to `useState`.

```suggestion
  const [existingNames, setExistingNames] = useState<FileName[]>([])
```

This removes the need for `@ts-expect-error` on `setExistingNames(existingNames)`.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts:27]: I'd suggest splitting this out into a custom hook (doesn't need to be a separate module) like `useExistingAnalyses` to make testing easier.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts:29]: Since this condition is checked in several places, a variable like `isGoogleWorkspace = !!googleProject` might be clearer.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts:94]: It looks like this module does not import `reportError` from src/libs/error.js. Thus, this is using [window.reportError](https://developer.mozilla.org/en-US/docs/Web/API/reportError). Adding the import / using the correct function removes the need for `@ts-expect-error`.

File: src/pages/workspaces/workspace/analysis/tool-utils.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/tool-utils.ts]: Formatting nit: this interface ends some lines with commas and some without.

File: src/pages/workspaces/workspace/analysis/tool-utils.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/tool-utils.ts]: Unlike other functions in this module, this function doesn't have a return type declared. TS can infer it, but it would be nice to have for documentation and for consistency.

File: src/pages/workspaces/workspace/workspace-utils.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/workspace-utils.ts]: Nit: "Info" seems superfluous. I'd suggest leaving it off and naming this "BaseWorkspace"

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/analysis/file-utils.ts]: I followed the pattern that `useWorkspaces` in `workspace-utils` does. Since `workspaceStore` is global, I think it makes sense for `analysisFilesStore` to be global since `analysisFiles` are used in multiple (3) components and are directly derived from the workspace in `workspaceStore`. Worth noting I kept this code in this PR since it is a file util, but it isn't actually used (or tested) until the subsequent PR (see current AnalysisDuplicator logic which calls the APIs that this component does directly). Could be convinced to move it out for now...

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/analysis/file-utils.ts]: Hmm... currently this is only (intended to be) used to conditionally display a spinner. Not sure what value it adds to make it a complex type over true/false? Open to pairing on this if you think it does add value.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts:22]: This PR just aims to move AnalysisDuplicator into its own file. That way, the follow-up PR displays all actual type-related changes needed.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts:27]: This is the custom hook in `file-utils`, but I'm avoiding any internal changes to this file as detailed above

File: src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts:29]: I use types for this in the follow-up PR

File: src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts:94]: I refactor this to use `withErrorReporting` in the follow-up PR

File: src/pages/workspaces/workspace/analysis/tool-utils.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/analysis/tool-utils.ts]: Ah, I'll add a linter rule

File: src/pages/workspaces/workspace/workspace-utils.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/workspace-utils.ts]: Hmm... I like this comment but in the other direction. Intention here is to differentiate between two variables often both named `workspace`, one of which is data nested within a parent object. I will rename the parent object as `BaseWorkspace` from `BaseWorkspaceWrapper`, but keep the child as `BaseWorkspaceInfo`.

File: .eslintrc.js
Comment (nawatts) [inline .eslintrc.js]: I'm in favor of this change, but since this requires changing existing code (DataTableProvider.ts) and ESLint changes affect all code, I think it is worth its own PR. @kyuksel since you've been working on coding guidelines, what do you think?

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (nawatts) [reply src/pages/workspaces/workspace/analysis/file-utils.ts]: I think this should be moved to src/libs/type-utils.

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (nawatts) [reply src/pages/workspaces/workspace/analysis/file-utils.ts]: Generally, state adds complexity and the larger in scope or the longer lived the state, the more complexity it adds. Since this hook refreshes on mount anyway, the only value that I see the global store providing is caching. However, because the store is global but the state is per-workspace, now we need to deal with resetting this when the workspace changes. As is, it looks like if a user looked at this in workspace A, then workspace B, they'd see analyses for workspace A in workspace B until the refresh request returned.

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (nawatts) [reply src/pages/workspaces/workspace/analysis/file-utils.ts]: There is some value in consistency, especially for reusable pieces. If a component uses two hooks that make network requests, it's easier for the developer if they both behave the same way. @msilva-broad has opinions on establishing patterns like this.

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (nawatts) [reply src/pages/workspaces/workspace/analysis/file-utils.ts]: Also, since this is more closely tied to refactoring the AnalysisDuplicator in a subsequent PR than the changes to notebook-utils that make up the bulk of this PR, I would have suggested putting in that PR instead of this one.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts
Comment (nawatts) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts:22]: Sounds good. In that case, I would have suggested leaving this as a JS file in this PR and converting it to TS in the follow up PR.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts]: Now that #3491 has merged, this should no longer be necessary.

File: src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts
Comment (nawatts) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts:94]: Though it will be refactored soon, the import needs to be added in this PR. Currently, this is using the wrong function and is introducing a bug.

File: src/pages/workspaces/workspace/workspace-utils.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/workspace-utils.ts]: Since the workspace object is used all over the place, these types should definitely live at a higher level than pages/workspaces/workspace. @msilva-broad do you have a preference on where?

**Review by nawatts — CHANGES_REQUESTED** [review]
Overall, I think the utility changes are looking pretty good.

Two things I think need to be changed:
- `reportError` in AnalysisDuplicator needs to be fixed
- Workspace types should be moved to a higher level since they'll be imported all over the place

I would suggest backing out the ESLint configuration change and the associated changes to other files, but I'll defer to @kyuksel on that.

There are a few things I think would have fit better in the following PR, but I won't request further changes over them.
- AnalysisDuplicator could have been left in place here or moved to a JS file to avoid adding `ts-expect-error` comments that will be resolved in the following PR.
- `useAnalysisFiles` could have been left to the PR that refactors AnalysisDuplicator. Since it's not used in this PR and there's still some unresolved discussion around it, it may be worth moving it to the next PR so that it doesn't hold up this one.

File: .eslintrc.js
Comment (kyuksel) [reply .eslintrc.js]: I appreciate the heads-up. 💯 on making separate PRs to isolate ESLint rule changes/additions. This makes reviews a lot easier since we know we don't expect any functional changes, and if something goes wrong, it makes dissecting potential culprits easier.

Comment (kyuksel) [reply .eslintrc.js]: P.S. I ❤️ the linting rule. Thanks for thinking of it @jdcanas 😊

File: src/pages/workspaces/workspace/workspace-utils.ts
Comment (msilva-broad) [reply src/pages/workspaces/workspace/workspace-utils.ts]: agreed.  If they are data-centric (which these appear to be), then the best spot would be within src/libs/ajax.

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (msilva-broad) [reply src/pages/workspaces/workspace/analysis/file-utils.ts]: to be fair, the LoadedState pattern was JUST added in recently.  But yes, I will be socializing it's broad adoption to promote consistency and separation-of-concerns.  Consistency will make the code-base easier to navigate and jump around in.

File: src/libs/ajax/file-browser-providers/GCSFileBrowserProvider.ts
Comment (msilva-broad) [inline src/libs/ajax/file-browser-providers/GCSFileBrowserProvider.ts]: declaring this type beforehand would be better I think.

File: .eslintrc.js
Comment (msilva-broad) [reply .eslintrc.js]: agreed.  This linter rule change also adds more noise to this PR.

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/analysis/file-utils.ts]: Yeah, I was starting to think the same. I'll remove it for now.

Comment (jdcanas) [reply src/pages/workspaces/workspace/analysis/file-utils.ts]: It should refresh on mount in any component that uses it... probably needs more testing in the subsequent PR though since I'm not super familiar with this `store` pattern and this is mostly copy/paste

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/analysis/file-utils.ts]: deferring this concern to follow-up duplicator PR. Sounds like its worth having a quick pairing session on this and other concerns about this hook

File: src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/analysis/modals/AnalysisDuplicator.ts:94]: ah, misunderstood the original comment. The fact there's an instance of `reportError` in scope by default made this confusing

File: src/pages/workspaces/workspace/workspace-utils.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/workspace-utils.ts]: moved it to ajax-common.js, rather than throwing more stuff in the larger Ajax module which we are trying to move away from

Comment (jdcanas) [reply src/pages/workspaces/workspace/workspace-utils.ts]: worth mentioning this suggestion does require changing ajax-common to a ts file...

File: .eslintrc.js
Comment (jdcanas) [reply .eslintrc.js]: Moved to this PR https://github.com/DataBiosphere/terra-ui/pull/3496. Easiest way to do this was to squash, soft reset, then re-commit the changes that did not have the lint rule noise. As such, PR chat history might be wonky, but I found this trade-off to be better than relying on me to transfer the changes to another branch rather than git.

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/file-utils.ts]: Should the example in the comment here be "file.txt" instead of "dir/file.txt"?

File: src/pages/workspaces/workspace/analysis/tool-utils.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/analysis/tool-utils.ts]: Naming nit: I think another name might read a bit more clearly. As is, this reads like "Is this tool label an app?". To me, it seems the answer to that question is always no: a label isn't an app.

Maybe one of `isAppToolLabel` or `isToolAnApp`?

File: src/pages/workspaces/workspace/workspace-utils.ts
Comment (nawatts) [inline src/pages/workspaces/workspace/workspace-utils.ts]: Since workspace is passed around and used by so many functions and components, I still think these types should live at a higher level than pages/workspaces/workspace.

**Review by nawatts — CHANGES_REQUESTED** [review]
A few nits, but only requesting changes over the placement of workspace types.

File: src/pages/workspaces/workspace/analysis/file-utils.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/analysis/file-utils.ts]: yeah, fixed. For some reason i think in a previous refactor that `getDisplayName` didn't actually strip the path, but it does in practice.

File: src/pages/workspaces/workspace/analysis/tool-utils.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/analysis/tool-utils.ts]: I like how `isAppToolLabel` reads

File: src/pages/workspaces/workspace/workspace-utils.ts
Comment (jdcanas) [reply src/pages/workspaces/workspace/workspace-utils.ts]: Hmm.. I thought you were talking about the cloud provider stuff originally but I see your point. I moved it to the same level the existing `workspace-utils.js` is, in a file called `workspace-utils.d.ts`. Is that the proper usage of a `d.ts` extension? It seems to work as intended and allow for slow migration from the `.js` definitions to `.ts`...

Comment (nawatts) [reply src/pages/workspaces/workspace/workspace-utils.ts]: Let's make it a plain `.ts` TypeScript file. [Declaration files](https://www.typescriptlang.org/docs/handbook/declaration-files/introduction.html) (`.d.ts` files) allow providing TypeScript types for code written in JS. Generally, `module.d.ts` would be paired with `module.js`. In this case, there's no `libs/workspace-utils.js` file (though there is a `components/workspace-utils.js`).

**Review by nawatts — APPROVED** [review]
I'd suggest moving `libs/workspace-utils.d.ts` to `libs/workspace-utils.ts`. Other than that, LGTM.

**Review by n8landolt — APPROVED** [review]

File: src/libs/ajax/file-browser-providers/GCSFileBrowserProvider.ts
Comment (msilva-broad) [reply src/libs/ajax/file-browser-providers/GCSFileBrowserProvider.ts]: I think this comment got lost in the PR shuffle.  This file looks to not have the comment addressed.  Inline types like this are harder to read.  It would be better to define RequestOptions type or interface just before this usage.

Comment (nawatts) [reply src/libs/ajax/file-browser-providers/GCSFileBrowserProvider.ts]: This file is no longer part of this PR. It was part of the ESLint configuration change that was split out into #3496.

**Review by msilva-broad — APPROVED** [review]
great work breaking up this work!

**Review by marctalbott — APPROVED** [review]

**Review by kyuksel — APPROVED** [review]

**Pull request merged**
